pandas/core/indexes/multi.py

```diff
@@ -689,7 +690,7 @@ def __len__(self) -> int:
     # --------------------------------------------------------------------
     # Levels Methods
 
-    @cache_readonly
+    @property
```

Comment:
the levels are actually not readonly, bacise each level's ``name`` attribute is writeable.

Reply:
Hmm I thought these were readonly as of 1.0 though?

https://pandas.pydata.org/pandas-docs/version/1.0.0/whatsnew/v1.0.0.html#backwards-incompatible-api-changes

@TomAugspurger

Reply:
I actually don't mind treating it as immutable, it was just that the test using ``pandas/tests/indexes/multi/test_names.py::check_level_names`` failed.

``CategoricalIndex`` and ``CategoricalDtype`` actually have the same issue already:

```python
>>> dtype = pd.CategoricalDtype(['a', 'b', 'c'])
>>> dtype.categories.name = "x"  # dtypes a re immutable, this is not supposed to work
>>> dtype.categories
Index(['a', 'b', 'c'], dtype='object', name='x')
```

IMO this issue is a wart, but nothing serious, because ``MultiIndex.names`` are now seperate from the levels, so it doesn't interfere with anything and the same issue is the same as with ``CategoricalIndex.categories.name``.

Reply:
Can you check the history here? Did we deliberately make this cache_readonly to avoid a performance regression.

I also don't see why `check_level_names` would be failing when this is cache_readonly.

Reply:
Yeah it was changed in #31651 because of a performance regression.

I've made a new version that does a shallow_copy of ``_cache["levels"]`` also.

Reply:
Using ``cache_readonly`` is faster than ``property``. I've updated the example to reflect that.